pkg/cmd/docs-issue-generation/github.go

```diff
@@ -0,0 +1,294 @@
+              messageHeadline
+              messageBody
+            }
+          }
+        }
+        pageInfo {
+          hasNextPage
+          endCursor
+        }
+      }
+    }
+  }
+}`
+	queryVariables := map[string]interface{}{
+		"prNumber": prNumber,
+	}
+	if cursor != "" {
+		queryVariables["cursor"] = cursor
+	}
+	err := queryGraphQL(
+		query,
+		queryVariables,
+		&search,
+	)
+	if err != nil {
+		fmt.Println(err)
+		return false, "", nil, err
+	}
+	for _, x := range search.Data.Repository.PullRequest.Commits.Edges {
+		matchingDocsIssue := prCommitsToExclude[prNumber][x.Node.Commit.Oid]
+		if nonBugFixRNRE.MatchString(x.Node.Commit.MessageHeadline) && matchingDocsIssue == "" {
+			commit := cockroachCommit{
+				Sha:             x.Node.Commit.Oid,
+				MessageHeadline: x.Node.Commit.MessageHeadline,
+				MessageBody:     x.Node.Commit.MessageHeadline,
+			}
+			result = append(result, commit)
+		}
+	}
+	pageInfo := search.Data.Repository.PullRequest.Commits.PageInfo
+	return pageInfo.HasNextPage, pageInfo.EndCursor, result, nil
+}
+
+var searchCockroachReleaseBranches = func() ([]string, error) {
+	var result []string
+	hasNextPage, nextCursor, branches, err := searchCockroachReleaseBranchesSingle("")
+	if err != nil {
+		fmt.Println(err)
+		return []string{}, err
+	}
+	result = append(result, branches...)
+	for hasNextPage {
+		hasNextPage, nextCursor, branches, err = searchCockroachReleaseBranchesSingle(nextCursor)
+		if err != nil {
+			fmt.Println(err)
+			return []string{}, err
+		}
+		result = append(result, branches...)
+	}
+	return result, nil
+}
+
+func searchCockroachReleaseBranchesSingle(cursor string) (bool, string, []string, error) {
+	var search gqlRef
+	var result []string
+	query := `query ($cursor: String) {
+  repository(owner: "cockroachdb", name: "cockroach") {
+    refs(
+      first: 100
+      refPrefix: "refs/"
+      query: "heads/release-"
+      after: $cursor
+      orderBy: {field: TAG_COMMIT_DATE, direction: DESC}
+    ) {
+      edges {
+        node {
+          name
+        }
+      }
+      pageInfo {
+        hasNextPage
+        endCursor
+      }
+    }
+  }
+}`
+	queryVariables := make(map[string]interface{})
+	if cursor != "" {
+		queryVariables["cursor"] = cursor
+	}
+	err := queryGraphQL(query, queryVariables, &search)
+	if err != nil {
+		fmt.Println(err)
+		return false, "", []string{}, err
+	}
+	for _, x := range search.Data.Repository.Refs.Edges {
+		result = append(result, strings.Replace(x.Node.Name, "heads/", "", -1))
+	}
+	pageInfo := search.Data.Repository.Refs.PageInfo
+	return pageInfo.HasNextPage, pageInfo.EndCursor, result, nil
+}
+
+// getJiraIssueFromGitHubIssue takes a GitHub issue and returns the appropriate Jira key from the issue body.
+// getJiraIssueFromGitHubIssue is specified as a function closure to allow for testing
+// of getJiraIssueFromGitHubIssue* methods.
+var getJiraIssueFromGitHubIssue = func(org, repo string, issue int) (string, error) {
```

Comment:
Nit: Bit weird to declare a top-level function like `var funcName = func(...)` Instead, I expect to see `func funcName(...)`.

Reply:
The reason for this is because I need to "mock" network resources for the tests to run. Take a look at like 769 in `docs_issue_generation_test.go`: https://github.com/cockroachdb/cockroach/pull/109788/files#diff-7e172e15a18c5d651442463a3286130095be65025f92e406a75727201a198dc9R769

If you know of a way that I can mock the network resources inside of this function without needing to declare the function as a variable, I'm all ears.

Reply:
Hmm. For now, it's fine. The "proper" way to do this would be to define an `interface` and put this function behind that interface. Then you have one instance for "production", and one for testing that has your "mock" logic.

No need to change anything for now and increase the complexity of what you have.

Reply:
Do you have an example in our codebase or elsewhere where that's done? I'll add that as a follow-up item for this ticket.

Reply:
Try an external resource like [this one](https://nathanleclaire.com/blog/2015/10/10/interfaces-and-composition-for-effective-unit-testing-in-golang/) or [this one](https://deliveroo.engineering/2019/05/17/testing-go-services-using-interfaces.html). It seems there are a ton of blog posts, if either of these don't make sense to you try Googling for them.

The short version is that this function would be "behind an interface" which might look like:

```
type jiraClient interface {
    getJiraIssueFromGitHubIssue(org, repo string, issue int) (string, error)
}

type productionJiraClient struct {}

func (_ productionJiraClient) getJiraIssueFromGitHubIssue(org, repo string, issue int) (string, error) {
   ....
}
```

The "production" `jiraClient` will just call `getJiraIssueFromGitHubIssue` as it's defined here. Any function that wants to call `getJiraIssueFromGitHubIssue()` needs a `jiraClient` instance. For testing, use a special instance that does whatever "mock" functionality you need.